File: modules/loopcam-recorder/android/src/main/java/expo/modules/loopcamrecorder/LoopcamRecorderModule.kt
package expo.modules.loopcamrecorder

import android.Manifest
import android.content.pm.PackageManager
import android.media.MediaMetadataRetriever
import android.os.Build
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import expo.modules.interfaces.permissions.PermissionsStatus
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.net.URI

/**
 * JS-facing surface of the recording engine (§3).
 *
 * Nothing here does timing-sensitive work: every call hands off to
 * [SegmentController], which owns the loop on its own thread. The bridge is a
 * control channel, not part of the recording path (§3.1).
 */
class LoopcamRecorderModule : Module(), SegmentController.Listener {

  private val context get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private val storage by lazy { StorageManager(context) }
  private val cameraRecorder by lazy { CameraXSegmentRecorder(context) }
  private val configStore by lazy { ConfigStore(context) }
  private val cameraProbe by lazy { CameraProbe(context) }

  override fun definition() = ModuleDefinition {
    Name("LoopcamRecorder")

    Events("onStateChange", "onClipFinished", "onSaved", "onStorageWarning", "onError")

    OnCreate {
      live = this@LoopcamRecorderModule
      // The tracker outlives every Activity — the foreground service keeps
      // recording with the app swiped away — so it is given the application
      // context here, once, rather than reaching for whatever is current.
      LocationTracker.attach(context)
      LocationTracker.configure(configStore.load())
      // §7.2 — a temp session that survived a process death is orphaned by
      // definition; sweep before anything else touches the buffer.
      storage.cleanupOrphanedSessions()
      // Kicked off here so the answer is ready by the time the first render
      // asks for it; the probe itself runs on its own thread.
      cameraProbe
    }

    OnDestroy {
      // A reload leaves the service and the camera bound to a controller that
      // no longer has anywhere to report; tear the whole thing down.
      if (live === this@LoopcamRecorderModule) live = null
      controller?.release()
      controller = null
      cameraRecorder.lifecycleOwner = null
      runCatching { RecordingService.stop(context) }
    }

    AsyncFunction("configure") { config: RecorderConfig ->
      // Persisted as well as applied. The service can outlive the JS side
      // (§5.1), so the stored copy is what a cold start reads — a config that
      // only lived in the controller would reset on every launch.
      configStore.save(config)
      requireController().configure(config)
    }

    /**
     * The live config if there is one, else whatever was last saved. Not the
     * type's defaults: those would tell a freshly-launched app it is set to the
     * back camera at 1080p regardless of what the user actually chose.
     */
    Function("getConfig") {
      (controller?.currentConfig() ?: configStore.load()).toMap()
    }

    Function("getCapabilities") {
      cameraProbe.capabilities()
    }

    AsyncFunction("requestPermissions") { promise: Promise ->
      // Only camera and mic gate recording. POST_NOTIFICATIONS is asked for at
      // the same time because the foreground service is invisible without it,
      // but a refusal must not block the drive. FOREGROUND_SERVICE_CAMERA is
      // install-time, not runtime — asking for it here would resolve false and
      // lock Play out entirely.
      val required = REQUIRED_PERMISSIONS
      // Asked at the same moment as camera and mic, and optional for the same
      // reason POST_NOTIFICATIONS is: a driver who refuses location still gets a
      // working dashcam, one whose footage stamps `--` where the speed would be.
      // Requesting it here rather than at Play also keeps the system dialog off
      // the screen of someone who has just started recording, quite possibly
      // while already moving.
      val optional = buildList {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
          add(Manifest.permission.POST_NOTIFICATIONS)
        }
        // Coarse alongside fine because Android 12+ shows the user a choice
        // between them, and asking for fine alone offers no approximate option
        // at all. A coarse-only grant is detected by LocationTracker and
        // reported through getLocationStatus rather than silently burning in
        // numbers from network positioning.
        add(Manifest.permission.ACCESS_FINE_LOCATION)
        add(Manifest.permission.ACCESS_COARSE_LOCATION)
      }.toTypedArray()
      val permissionsManager = appContext.permissions
      if (permissionsManager == null) {
        promise.resolve(false)
      } else {
        permissionsManager.askForPermissions({ result ->
          promise.resolve(required.all { result[it]?.status == PermissionsStatus.GRANTED })
        }, *(required + optional))
      }
    }

    /**
     * Whether recording could start right now without a prompt. Synchronous and
     * side-effect free on purpose: it is asked on mount to decide whether the
     * viewfinder may light up, and a screen opening is not the moment to throw a
     * system dialog at someone who has not asked for anything yet.
     */
    Function("hasPermissions") {
      REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
      }
    }

    /**
     * Why the burned-in speed is reading `--`, for Settings to explain.
     *
     * A permanently blank speed field has several quite different causes — no
     * lock yet, a tunnel, a refused permission, an approximate-only grant — and
     * on a screen the driver only looks at while parked, guessing which is not
     * something to leave to them.
     */
    Function("getLocationStatus") {
      LocationTracker.status().jsValue
    }

    /**
     * Light the viewfinder without recording. No foreground service and no
     * files — this is the picture only, bound to the Activity so it goes away
     * with the screen.
     */
    AsyncFunction("startPreview") {
      val owner = appContext.currentActivity as? LifecycleOwner
        ?: throw Exceptions.MissingActivity()
      // The stored mode, so the standby picture is laid out the way a recording
      // would be — including the front-camera inset — rather than only
      // revealing the layout once Play is pressed.
      val mode = (controller?.currentConfig() ?: configStore.load()).camera
      cameraRecorder.startPreview(owner, mode)
    }

    AsyncFunction("stopPreview") {
      cameraRecorder.stopPreview()
    }

    /**
     * PLAY — the foreground service must be up *before* the camera binds, since
     * it is the LifecycleOwner that keeps the session alive in the background
     * (§5.1).
     */
    AsyncFunction("start") { promise: Promise ->
      RecordingService.startAndAwait(context) { service ->
        if (service == null) {
          promise.reject(
            "ERR_SERVICE_START",
            "The DashCam recording service did not start",
            null,
          )
          return@startAndAwait
        }
        cameraRecorder.lifecycleOwner = service
        requireController().start { result ->
          result
            .onSuccess { promise.resolve(it.toMap()) }
            .onFailure { promise.reject("ERR_START_FAILED", it.message, it) }
        }
      }
    }

    /** STOP. */
    AsyncFunction("stop") { promise: Promise ->
      val segmentController = controller
      if (segmentController == null) {
        promise.resolve(idleStatus().toMap())
        return@AsyncFunction
      }
      segmentController.stop { status ->
        cameraRecorder.lifecycleOwner = null
        RecordingService.stop(context)
        promise.resolve(status.toMap())
      }
    }

    /** SAVE — resolves once the merged file is on disk; recording never pauses. */
    AsyncFunction("save") { promise: Promise ->
      requireController().save(SaveTrigger.MANUAL) { result ->
        result
          .onSuccess { promise.resolve(it.toMap()) }
          .onFailure { promise.reject("ERR_SAVE_FAILED", it.message, it) }
      }
    }

    Function("getStatus") {
      (controller?.status() ?: idleStatus()).toMap()
    }

    AsyncFunction("listSavedClips") {
      savedClips().map { it.toMap() }
    }

    AsyncFunction("deleteSavedClip") { id: String ->
      savedClips().firstOrNull { it.id == id }?.let { clip ->
        val video = File(URI(clip.uri))
        // The marker goes with its clip; an orphan would silently protect the
        // next clip that happened to land on the same timestamped name.
        storage.protectionMarkerFor(video).delete()
        video.delete()
        clip.metadataUri?.let { File(URI(it)).delete() }
      }
    }

    AsyncFunction("setClipProtected") { id: String, isProtected: Boolean ->
      savedClips().firstOrNull { it.id == id }?.let { clip ->
        storage.setProtected(File(URI(clip.uri)), isProtected)
      }
    }

    AsyncFunction("getStorageStatus") {
      val (count, bytes) = storage.savedFootprint()
      storage.storageStatus(count, bytes).toMap()
    }

    AsyncFunction("cleanupOrphanedClips") {
      storage.cleanupOrphanedSessions()
    }

    View(LoopcamRecorderView::class) {
      // No `lens` prop: which camera records is `RecorderConfig.cameraMode`,
      // owned by the controller. A view prop that also selected it could only
      // ever be the second, disagreeing source of truth.
      Prop("resizeMode") { view: LoopcamRecorderView, mode: String ->
        view.setResizeMode(mode)
      }
    }
  }

  // --- SegmentController.Listener ----------------------------------------

  override fun onStateChanged(status: BufferStatus) = sendEvent("onStateChange", status.toMap())

  override fun onClipFinished(status: BufferStatus) {
    // The notification is the only view of the buffer once the app is
    // backgrounded, and a clip boundary is the cheapest honest moment to
    // refresh it (§6).
    RecordingService.updateNotification(status.bufferedSec)
    sendEvent("onClipFinished", status.toMap())
  }

  override fun onSaved(clip: SavedClip) {
    sendEvent("onSaved", clip.toMap())

    // A save is the only moment saved storage grows, so the only moment the
    // budget can be breached (§7.2).
    val deleted = storage.enforceBudget()
    val (count, bytes) = storage.savedFootprint()
    val status = storage.storageStatus(count, bytes)
    if (deleted.isNotEmpty() || status.lowSpaceWarning) {
      sendEvent("onStorageWarning", status.toMap() + mapOf("deletedClipIds" to deleted))
    }
  }

  override fun onError(code: RecorderErrorCode, message: String) =
    sendEvent("onError", mapOf("code" to code.jsValue, "message" to message))

  // --- notification controls ----------------------------------------------

  /**
   * The lock-screen Stop, running the same teardown as the JS `stop` — the
   * camera's LifecycleOwner has to be dropped here too, or the next Play binds
   * CameraX to a service that is on its way to being destroyed.
   *
   * [onDone] fires only once the controller has actually finished, so the
   * caller knows when it is safe to let the service die.
   */
  internal fun stopFromNotification(onDone: () -> Unit) {
    val segmentController = controller
    if (segmentController == null) {
      cameraRecorder.lifecycleOwner = null
      onDone()
      return
    }
    segmentController.stop {
      cameraRecorder.lifecycleOwner = null
      onDone()
    }
  }

  /** The lock-screen Save. JS still hears about it through `onSaved` (§3.1). */
  internal fun saveFromNotification(onResult: (Result<SavedClip>) -> Unit) {
    val segmentController = controller
    if (segmentController == null) {
      onResult(Result.failure(IllegalStateException("Not recording")))
      return
    }
    segmentController.save(SaveTrigger.MANUAL, onResult)
  }

  // --- wiring -------------------------------------------------------------

  /**
   * The controller needs no Activity: the camera's LifecycleOwner is
   * [RecordingService], handed to the recorder at Play. That also means
   * `configure` works before the first Play, which it could not when this
   * required a live Activity.
   */
  private fun requireController(): SegmentController {
    controller?.let { return it }
    return SegmentController(
      storage = storage,
      recorder = cameraRecorder,
      merger = ClipMerger(storage),
      listener = this,
    ).also {
      // Seeded from disk, not from the type's defaults: a Play issued before JS
      // has pushed anything — the notification's action, or a restart — must
      // record with the camera and tier the user actually chose.
      it.configure(configStore.load())
      controller = it
    }
  }

  private fun idleStatus(): BufferStatus {
    val config = controller?.currentConfig() ?: configStore.load()
    return BufferStatus(
      state = RecorderState.IDLE,
      clipCount = 0,
      maxClips = config.maxClips,
      bufferedSec = 0.0,
      bufferedBytes = 0L,
      elapsedSec = 0.0,
    )
  }

  /**
   * §7.1 — the saved directory *is* the index: one .mp4 plus an optional .json
   * sidecar per incident. No database to fall out of sync with disk.
   */
  private fun savedClips(): List<SavedClip> {
    // Empty files are skipped rather than listed: a merge interrupted by a
    // process death can still leave one, and an unplayable row in the gallery
    // is worse than a missing one.
    val files = storage.savedRoot.listFiles { f: File -> f.extension == "mp4" && f.length() > 0L }
      ?: return emptyList()
    return files.sortedByDescending { it.lastModified() }.map { file ->
      val sidecar = storage.metadataFileFor(file)
      SavedClip(
        id = file.nameWithoutExtension,
        uri = file.toURI().toString(),
        metadataUri = if (sidecar.exists()) sidecar.toURI().toString() else null,
        createdAtMs = file.lastModified(),
        durationSec = durationSecOf(file),
        sizeBytes = file.length(),
        isProtected = storage.isProtected(file),
        trigger = SaveTrigger.MANUAL,
      )
    }
  }

  /**
   * The container is the only honest source for how long a saved clip runs: the
   * merge concatenates a variable number of segments, and the last one is
   * usually cut short by the Save itself, so no arithmetic over the config
   * predicts it. A file that cannot be read reports 0 rather than guessing.
   */
  private fun durationSecOf(file: File): Double {
    val retriever = MediaMetadataRetriever()
    return try {
      retriever.setDataSource(file.absolutePath)
      val ms = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull()
      if (ms == null) 0.0 else ms / 1000.0
    } catch (t: Throwable) {
      Log.w(TAG, "Could not read duration from ${file.name}", t)
      0.0
    } finally {
      runCatching { retriever.release() }
    }
  }

  companion object {
    private const val TAG = "LoopCam/Module"

    /** The two that actually gate capture; everything else is a nicety. */
    private val REQUIRED_PERMISSIONS =
      arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO)

    /** Shared with [RecordingService] so notification actions hit the live loop. */
    @JvmStatic
    var controller: SegmentController? = null
      internal set

    /**
     * The attached module, so [RecordingService] can route a notification tap
     * through the same code the JS buttons use rather than reaching past it
     * into the controller. Null between a reload and the next OnCreate.
     */
    @JvmStatic
    internal var live: LoopcamRecorderModule? = null
  }
}

File: modules/loopcam-recorder/ios/LoopcamRecorderModule.swift
import AVFoundation
import ExpoModulesCore
import UIKit

/// JS-facing surface of the recording engine (§3).
///
/// Nothing here does timing-sensitive work: every call hands off to
/// `SegmentController`, which owns the loop on its own queue. The bridge is a
/// control channel, not part of the recording path (§3.1).
public class LoopcamRecorderModule: Module, SegmentControllerDelegate {
  private let storage = StorageManager()
  private lazy var recorder = AVSegmentRecorder()
  private lazy var controller = SegmentController(
    storage: storage,
    recorder: recorder,
    merger: ClipMerger(storage: storage),
    delegate: self
  )
  private var protectedIds = Set<String>()
  /// Live Activity button taps. Held so `OnDestroy` can drop them — a reload
  /// otherwise leaves a dead module answering the Lock Screen.
  private var liveActivityObservers: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("LoopcamRecorder")

    Events("onStateChange", "onClipFinished", "onSaved", "onStorageWarning", "onError")

    OnCreate {
      // §7.2 — a temp session that survived a crash is orphaned by definition.
      self.storage.cleanupOrphanedSessions()
      // §6 — needed for the low-battery auto-save-and-stop threshold.
      UIDevice.current.isBatteryMonitoringEnabled = true
      self.observeLiveActivity()
    }

    OnDestroy {
      self.liveActivityObservers.forEach(NotificationCenter.default.removeObserver)
      self.liveActivityObservers = []
      // The session dies with the module, so the card has to go with it rather
      // than sit on the Lock Screen offering buttons that no longer land.
      LiveActivityBridge.postEnd()
      self.controller.release()
    }

    AsyncFunction("configure") { (config: RecorderConfig) in
      self.controller.configure(config)
    }

    Function("getConfig") {
      self.controller.currentConfig.asDictionary()
    }

    AsyncFunction("requestPermissions") { (promise: Promise) in
      // TODO(phase-2): also request CoreLocation "when in use" if GPS tagging
      // is on, and surface the §5.2 "keep the screen on while driving" copy.
      AVCaptureDevice.requestAccess(for: .video) { videoGranted in
        AVCaptureDevice.requestAccess(for: .audio) { audioGranted in
          promise.resolve(videoGranted && audioGranted)
        }
      }
    }

    /// PLAY.
    AsyncFunction("start") { () -> [String: Any] in
      // §5.2 — iOS has no foreground-service equivalent. Keeping the screen
      // awake is the honest mechanism behind "driving mode"; an active audio
      // session (configured in AVSegmentRecorder) is what buys background time.
      DispatchQueue.main.async { UIApplication.shared.isIdleTimerDisabled = true }
      self.controller.start()
      // The Lock Screen card is the only control surface once the phone locks.
      LiveActivityBridge.post(LiveActivityBridge.start, self.controller.status)
      return self.controller.status.asDictionary()
    }

    /// STOP.
    AsyncFunction("stop") { () -> [String: Any] in
      self.controller.stop()
      DispatchQueue.main.async { UIApplication.shared.isIdleTimerDisabled = false }
      LiveActivityBridge.postEnd()
      return self.controller.status.asDictionary()
    }

    /// SAVE — resolves once the merged file is on disk; recording never pauses.
    AsyncFunction("save") { (promise: Promise) in
      self.controller.save(trigger: .manual) { result in
        switch result {
        case .success(let clip):
          promise.resolve(clip.asDictionary())
        case .failure(let error):
          promise.reject("ERR_SAVE_FAILED", error.localizedDescription)
        }
      }
    }

    Function("getStatus") {
      self.controller.status.asDictionary()
    }

    AsyncFunction("listSavedClips") { () -> [[String: Any?]] in
      self.savedClips().map { $0.asDictionary() }
    }

    AsyncFunction("deleteSavedClip") { (id: String) in
      guard let clip = self.savedClips().first(where: { $0.id == id }) else { return }
      try? FileManager.default.removeItem(at: clip.url)
      if let metadataURL = clip.metadataURL {
        try? FileManager.default.removeItem(at: metadataURL)
      }
    }

    AsyncFunction("setClipProtected") { (id: String, isProtected: Bool) in
      // TODO(phase-4): persist alongside the sidecar so it survives a restart.
      if isProtected {
        self.protectedIds.insert(id)
      } else {
        self.protectedIds.remove(id)
      }
    }

    AsyncFunction("getStorageStatus") { () -> [String: Any] in
      let clips = self.savedClips()
      return self.storage.storageStatus(
        savedClipCount: clips.count,
        savedBytes: clips.reduce(0) { $0 + $1.sizeBytes }
      ).asDictionary()
    }

    AsyncFunction("cleanupOrphanedClips") { () -> Int in
      self.storage.cleanupOrphanedSessions()
    }

    View(LoopcamRecorderView.self) {
      Prop("lens") { (view: LoopcamRecorderView, lens: String) in
        view.setLens(lens)
      }
      Prop("resizeMode") { (view: LoopcamRecorderView, mode: String) in
        view.setResizeMode(mode)
      }
    }
  }

  // MARK: - SegmentControllerDelegate

  func segmentControllerDidChangeState(_ status: BufferStatus) {
    sendEvent("onStateChange", status.asDictionary())
    if status.state != .idle {
      LiveActivityBridge.post(LiveActivityBridge.update, status)
    }
  }

  /// The card is repainted here rather than on a timer: the buffer figure only
  /// changes when a clip closes (§2.4), and a Live Activity update per second
  /// would burn the system's update budget for a number that had not moved.
  func segmentControllerDidFinishClip(_ status: BufferStatus) {
    sendEvent("onClipFinished", status.asDictionary())
    LiveActivityBridge.post(LiveActivityBridge.update, status)
  }

  func segmentControllerDidSave(_ clip: SavedClip) {
    sendEvent("onSaved", clip.asDictionary() as [String: Any])
    LiveActivityBridge.post(LiveActivityBridge.update, controller.status, banner: "Clip saved")
  }

  func segmentControllerDidError(_ code: RecorderErrorCode, _ message: String) {
    sendEvent("onError", ["code": code.rawValue, "message": message])
    LiveActivityBridge.post(LiveActivityBridge.update, controller.status, banner: message)
  }

  // MARK: - Lock Screen controls

  /// Both handlers mirror the Android notification actions: they run the *same*
  /// engine calls as the JS buttons, so a Save from the Lock Screen is
  /// indistinguishable downstream — JS still hears `onSaved` (§3.1).
  private func observeLiveActivity() {
    let center = NotificationCenter.default
    liveActivityObservers = [
      center.addObserver(forName: LiveActivityBridge.saveRequested, object: nil, queue: nil) { [weak self] _ in
        self?.saveFromLiveActivity()
      },
      center.addObserver(forName: LiveActivityBridge.stopRequested, object: nil, queue: nil) { [weak self] _ in
        self?.stopFromLiveActivity()
      },
    ]
  }

  /// Save cuts the in-flight clip and waits for the merge (§2.3), which can take
  /// until the next clip boundary. With the phone locked the card is the only
  /// feedback there is, so it has to say something immediately — otherwise a
  /// working Save is indistinguishable from a dead button.
  private func saveFromLiveActivity() {
    LiveActivityBridge.post(LiveActivityBridge.update, controller.status, banner: "Saving…")
    controller.save(trigger: .manual) { [weak self] result in
      guard let self else { return }
      switch result {
      case .success:
        // `segmentControllerDidSave` already repainted the card with the
        // "Clip saved" banner, and JS was told. Nothing left to do here.
        break
      case .failure(let error):
        LiveActivityBridge.post(
          LiveActivityBridge.update,
          self.controller.status,
          banner: error.localizedDescription
        )
      }
    }
  }

  private func stopFromLiveActivity() {
    LiveActivityBridge.post(LiveActivityBridge.update, controller.status, banner: "Stopping…")
    controller.stop()
    DispatchQueue.main.async { UIApplication.shared.isIdleTimerDisabled = false }
    LiveActivityBridge.postEnd()
  }

  // MARK: - saved clips

  /// §7.1 — the saved directory *is* the index: one .mp4 plus an optional .json
  /// sidecar per incident. No database to fall out of sync with disk.
  private func savedClips() -> [SavedClip] {
    let fm = FileManager.default
    guard let urls = try? fm.contentsOfDirectory(
      at: storage.savedRoot,
      includingPropertiesForKeys: [.contentModificationDateKey]
    ) else { return [] }

    return urls
      .filter { $0.pathExtension == "mp4" }
      .sorted { lhs, rhs in modifiedAt(lhs) > modifiedAt(rhs) }
      .map { url in
        let sidecar = storage.metadataURL(for: url)
        let id = url.deletingPathExtension().lastPathComponent
        return SavedClip(
          id: id,
          url: url,
          metadataURL: fm.fileExists(atPath: sidecar.path) ? sidecar : nil,
          createdAtMs: modifiedAt(url).timeIntervalSince1970 * 1000,
          // TODO(phase-3): read the real duration from AVAsset.
          durationSec: 0,
          sizeBytes: storage.fileSize(at: url),
          isProtected: protectedIds.contains(id),
          trigger: .manual
        )
      }
  }

  private func modifiedAt(_ url: URL) -> Date {
    (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate)
      ?? Date.distantPast
  }
}
